Libraries/oneTBB/tbb-resumable-tasks-sycl/src/tbb-resumable-tasks-sycl.cpp

```diff
@@ -0,0 +1,137 @@
+//==============================================================
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+
+#include <cmath>
+#include <array>
+#include <thread>
+#include <iostream>
+#include <atomic>
+#include <sstream>
+
+#include <CL/sycl.hpp>
+
+#include <tbb/blocked_range.h>
+#include <tbb/global_control.h>
+#include <tbb/task.h>
+#include <tbb/task_group.h>
+#include <tbb/parallel_for.h>
+
+constexpr cl::sycl::access::mode sycl_read = cl::sycl::access::mode::read;
+constexpr cl::sycl::access::mode sycl_write = cl::sycl::access::mode::write;
+
+const float ratio = 0.5; // CPU or GPU offload ratio
+const float alpha = 0.5; // coeff for triad calculation
+
+constexpr std::size_t array_size = 16;
+std::array<float, array_size> a_array; // input array
+std::array<float, array_size> b_array; // input array
+std::array<float, array_size> c_array; // output array
+
+void print_array( const char* text, const std::array<float, array_size>& array ) {
+    std::cout << text;
+    for (const auto& s: array) std::cout << s << ' ';
+    std::cout << std::endl;
+}
+
+class AsyncActivity {
+    float offload_ratio;
+    std::atomic<bool> submit_flag;
+    tbb::task::suspend_point suspend_point;
+    std::thread service_thread;
+
+public:
+    AsyncActivity() : offload_ratio(0), submit_flag(false),
+        service_thread([this] {
+            while(!submit_flag) {
+                std::this_thread::yield();
+            }
+            std::size_t array_size_sycl = std::ceil(array_size * offload_ratio);
+            std::stringstream sstream;

```

Comment:
Why?

Reply:
The reason is that AsyncActivity thread lambda and task in the task_group (line 100) are executed simultaneously. If we use std::cout without stringstream, the output can be mixed.

Reply:
So, why this fundamental design point does not appear proudly in the comments?
This is a **TUTORIAL**, not some random write-only code for some obfuscation context trying to waste a maximum of time of the code reviewers... :-(
PLEASE USE A LOT OF COMMENTS IN THIS REPOSITORY, like any collaborative open-source project in general.